### PR TITLE
provide a helm chart for sechub server #1567

### LINKED ISSUE #1567: Provide Helm Chart for SecHub
Provide a Helm Chart for SecHub. The helm chart will allow one to run SecHub in a Kubernetes cluster.

Part of: #320

### PR DESCRIPTION
**Pull request by sven-dmlr**
closes #1567

### CODE REVIEW

File: sechub-solution/helm/sechub-server/values.yaml [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+  baseURL: https://to.be.defined
+  # Put the user id  of Checkmarx user in k8s secret secret-checkmarx:sechub_adapter_checkmarx_userid
+  # Put the password of Checkmarx user in k8s secret secret-checkmarx:sechub_adapter_checkmarx_password
+  #
+  # Interval in minutes when SecHub checks if the scan has finished on Checkmarx side
+  checkIntervalInMinutes: 1
+  # Time in minutes when adapter result check will be canceled/time out. When -1 timeout is 7200 minutes
+  timeoutInMinutes: 600
+  # Team ID of a newly created Checkmarx project.
+  # For project creation, a team must be defined.
```
Comment (Jeeppler) [inline sechub-solution/helm/sechub-server/values.yaml:188]:
```suggestion
  # For project creation, a Checkmarx team id must be defined.
```

**Review by Jeeppler — CHANGES_REQUESTED** [review]
See comments.

File: sechub-solution/helm/sechub-server/LICENSE [hunk 2]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2021 Daimler TSS GmbH
```
[the remark this answers is not in the document]
Comment (sven-dmlr) [reply sechub-solution/helm/sechub-server/LICENSE:3]: Changed.
Weather to change all others needs to be discussed in the core team.

File: sechub-solution/helm/sechub-server/templates/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+{{- end}}
+
+# ----------------------------------------------------------------------------------------------#
+# - Scope: system
+# --------------------------------------------------------------------------------------------- #
+        # Maximum memory usage percentage accepted by sechub system. Can be a value from 50 up to 100 for 100%
+        - name: SECHUB_MONITORING_ACCEPTED_MEMORY_USAGE_MAX
+          value: "{{ .Values.memory.maxPercentage }}"
+# ----------------------------------------------------------------------------------------------#
+# - Scope: target
```
[the remark this answers is not in the document]
Comment (sven-dmlr) [reply sechub-solution/helm/sechub-server/templates/deployment.yaml:536]: done

File: sechub-solution/helm/sechub-server/templates/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+# --------------------------------------------------------------------------------------------- #
+        # *Strategy to decide intranet/internet targets by given URI.*
+        # Starts always with strategy-identifer, colon and value(s). Currently only 'intranet-hostname-ends-with' is supported as strategy.
+        # For example: `intranet-hostname-ends-with::intranet.example.org,intx.example.com`.
+        # All other hostnames are considered as INTERNET hosts.
+        # But no matter if strategy is defined or not: loopback addresses are always ignored.
+        - name: SECHUB_TARGET_RESOLVE_STRATEGY_URI
+          value: "{{ .Values.webScan.intranetDetection.uriPattern }}"
+        # *Strategy to decide intranet/internet targets by given IP.*
```
[the remark this answers is not in the document]
Comment (sven-dmlr) [reply sechub-solution/helm/sechub-server/templates/deployment.yaml:545]: done

File: sechub-solution/helm/sechub-server/templates/deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,555 @@
...
+        # Starts always with strategy-identifer, colon and value(s). Currently only 'intranet-ip-pattern' is supported as strategy.
+        # For example: `intranet-ip-pattern:192.168.178.*,[2001:db8:85a3:0:0:8a2e:370:*]`.
+        # All other hostnames are considered as INTERNET hosts.
+        # But no matter if strategy is defined or not: loopback addresses are always ignored.
```
[the remark this answers is not in the document]
Comment (sven-dmlr) [reply sechub-solution/helm/sechub-server/templates/deployment.yaml:549]: done

File: sechub-solution/helm/sechub-server/values.yaml [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+  # imagePullSecrets: |-
+  #   - name: sechub-pull-secret
+  #   - name: sps2
+
+name: sechub-server
+
+memory:
+  # Memory for the SecHub server container (increase to e.g. 3Gi for larger installations)
+  container: 1Gi
+  # Maximum memory usage percentage accepted by sechub system. Can be a value from 50 up to 100 for 100%
```
[the remark this answers is not in the document]
Comment (sven-dmlr) [reply sechub-solution/helm/sechub-server/values.yaml:23]: done

File: sechub-solution/helm/sechub-server/values.yaml [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+server:
+  # email address of the SecHub administrators. Informations about scheduler stop/start, new users etc. will go there.
+  adminEmail: mail@to.be.defined
+  baseURL:
+    # Base url for SecHub administration server instance (with Spring profile `admin_access`)
+    adminAccess: https://to.be.defined
+    # SecHub base url for users. Can also be an ip address like https://10.10.10.10
+    userAccess: https://to.be.defined
+  # db connection - can be omitted when using spring profile `h2` (in-memory db)
+  databaseConnection: "jdbc:postgresql://postgres-db:5432/sechub"
```
[the remark this answers is not in the document]
Comment (sven-dmlr) [reply sechub-solution/helm/sechub-server/values.yaml:35]: I will leave it as it is.
I've added a warning hint:
```
# db connection - can be omitted when using spring profile `h2` (in-memory db)
# ! But caution: You can only use `h2` with _one_ SecHub server (replicaCount=1)!
# !              As soon as you deploy a cluster, you must set up a shared database.
```

The default Spring profiles are: `prod,admin_access`. `prod` implies a Postgres database.

I've also added a comment about how the database credentials have to be defined.

File: sechub-solution/helm/sechub-server/values.yaml (hunk 1 above)
Comment (sven-dmlr) [reply sechub-solution/helm/sechub-server/values.yaml:188]: Fixed. (The preceding comment line contained basically the same informations)

File: sechub-solution/helm/sechub-server/templates/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,555 @@
...
+        - name: SECHUB_NOTIFICATION_SMTP_HOSTNAME
+          value: "{{ .Values.server.email.smtp.host }}"
+        # Port of SMTP server
+        - name: SECHUB_NOTIFICATION_SMTP_PORT
+          value: "{{ .Values.server.email.smtp.port }}"
+# ----------------------------------------------------------------------------------------------#
+# - Scope: scheduler
+# --------------------------------------------------------------------------------------------- #
+        # Job scheduling is activated in intervals
+        # To avoid schedulers running always on same time in a cluster, we also have
```
[the remark this answers is not in the document]
Comment (sven-dmlr) [reply sechub-solution/helm/sechub-server/templates/deployment.yaml:452]: The comment about  SECHUB_CONFIG_TRIGGER_NEXTJOB_INITIALDELAY was confusing.
Changed the comment so it describes SECHUB_CONFIG_TRIGGER_NEXTJOB_DELAY.

File: sechub-solution/helm/sechub-server/templates/deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,555 @@
...
+# --------------------------------------------------------------------------------------------- #
+        # Base url for administration server - e.g. https://sechub.example.com:8443
+        - name: SECHUB_SERVER_ADMINISTRATION_BASEURL
+          value: "{{ .Values.server.baseURL.adminAccess }}"
+        # NPM for adminstrator emails
+        - name: SECHUB_NOTIFICATION_EMAIL_ADMINISTRATORS
+          value: "{{ .Values.server.adminEmail }}"
+{{- if .Values.adapterCheckmarx.enabled }}
+# ----------------------------------------------------------------------------------------------#
+# - Scope: Checkmarx
```
[the remark this answers is not in the document]
Comment (sven-dmlr) [reply sechub-solution/helm/sechub-server/templates/deployment.yaml:186]: Added comments here and at the PDS-Checkmarx section.

File: sechub-solution/helm/sechub-server/templates/deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,555 @@
...
+          value: "{{ .Values.server.spring.hikariPool.size.minimumIdle }}"
+        - name: SPRING_DATASOURCE_HIKARI_MAXIMUMPOOLSIZE
+          value: "{{ .Values.server.spring.hikariPool.size.maximum | toString }}"
+# ----------------------------------------------------------------------------------------------#
+# - Scope: administration
+# --------------------------------------------------------------------------------------------- #
+        # Base url for administration server - e.g. https://sechub.example.com:8443
+        - name: SECHUB_SERVER_ADMINISTRATION_BASEURL
+          value: "{{ .Values.server.baseURL.adminAccess }}"
+        # NPM for adminstrator emails
```
[the remark this answers is not in the document]
Comment (sven-dmlr) [reply sechub-solution/helm/sechub-server/templates/deployment.yaml:181]: npm: non-personal mailbox was meant :-D
Duplicated the comment from the values.yaml file, so we have the same comment here.

File: sechub-solution/helm/sechub-server/templates/deployment.yaml [hunk 11]
```diff
@@ -0,0 +1,555 @@
...
+          httpGet:
+            scheme: HTTPS
+            path: /api/anonymous/check/alive
+            port: sechub-https
+          failureThreshold: 3
+          periodSeconds: 10
+          successThreshold: 1
+          timeoutSeconds: 3
+        volumeMounts:
+          - mountPath: /home/javarun/secret-ssl
```
[the remark this answers is not in the document]
Comment (sven-dmlr) [reply sechub-solution/helm/sechub-server/templates/deployment.yaml:82]: Changed to `/sechub/secrets/secret-ssl`

File: sechub-solution/helm/sechub-server/templates/deployment.yaml [hunk 12]
```diff
@@ -0,0 +1,555 @@
...
+      securityContext:
+        runAsUser: {{ .Values.runAs.uid }}  # user id of the application user. (overrides setings in the Dockerfile)
+        runAsGroup: {{ .Values.runAs.gid }} # group id of the application group. (overrides setings in the Dockerfile)
+        fsGroup: {{ .Values.runAs.gid }}    # group id of the application group. This group has write access to the volumes.
+      volumes:
+        - name: secret-ssl
+          secret:
+            secretName: secret-ssl
+        - name: tmp-volume
+          ephemeral:
```
[the remark this answers is not in the document]
Comment (sven-dmlr) [reply sechub-solution/helm/sechub-server/templates/deployment.yaml:41]: As discussed:
- tmp-volume is now configurable (enable/disable)
- when enabled, the type can be chosen between emptyDir and genericEphemeral

**Review by Jeeppler — APPROVED** [review]
Looks good.

**Pull request merged**